[FLINK-16086][chinese-translation]Translate "Temporal Tables" page of "Streaming Concepts" into Chinese (#11978)
apache/flink: docs/dev/table/streaming/temporal_tables.zh.md

---

> title: "时态表（Temporal Tables）"

Reviewer: “时态表” 这个偏专有名词，需要其他人确认下是否合理。

under the License.
-->

> 时态表（Temporal Table）表示一直在变化的表上一个（参数化）视图的概念，该视图返回表在某个特定时间点的内容。

Reviewer: “时态表（Temporal Table）表示一直在变化的表上一个（参数化）视图的概念” 这句话读起来有点点拗口

> 时态表（Temporal Table）表示基于记录变更历史的表上（参数化）视图的概念，该视图返回表在某个特定时间点的内容。

Reviewer: `时态表（Temporal Table）表示基于记录变更历史的表上（参数化）视图的概念` 这句话读起来有点拗口，有办法优化一下吗？

> 变化的表可以是追踪变化的变化历史记录表（例如：数据库变更日志），也可以是有具体更改的维表（例如：数据库表）。

Reviewer: “变化的表可以是追踪变化的变化历史记录表（例如：数据库变更日志）” -> “变更表可以是跟踪变化的历史记录表（如数据库变化记录）。” 会更好一点吗？

> 对于记录变更历史的表，Flink 可以追踪这些变化，并且允许查询这张表在某个特定时间点的内容。在 Flink 中，这类表由 *时态表函数（Temporal Table Function）* 表示。

Reviewer: ```suggestion
对于记录变更历史的表，Flink 可以追踪这些变化，并且允许查询这张表在某个特定时间点的内容。在 Flink 中，这类表由*时态表函数（Temporal Table Function）* 表示。
```

> 对于表的历史变化，Flink 可以追踪这些变化，并且允许查询这张表在某个特定时间点的内容。在 Flink 中，这类表由 *时态表函数（Temporal Table Function）* 表示。

Reviewer: 这里的 “changing history table”指上面两种的一种吧，表示“changelog”。这里是否有比 “对于表的历史变化”更好的翻译呢？

Author: 多谢review！“changing history table” 我理解实际上是指 记录事件变更历史的表，最终如何翻译还要再讨论下。

对于变化的维表，Flink 允许查询这张表在处理时的内容，在 Flink 中，此类表由 *时态表（Temporal Table）* 表示。

* This will be replaced by the TOC
{:toc}

设计初衷
----------

> ### 与表的历史变化相关

Reviewer: “changing history table” 有一个更好的翻译吗？现在的翻译不能体现“changelog”

> 假设我们有下表 `RatesHistory`。

Reviewer: 个人观点：这里翻译成 ”假设我们有表 `RateHistory` 如下所示“ 会好一些吗？现在的翻译感觉句子没有说完一样的。当然有其他更好的翻译更好。

11:49   Pounds      108
{% endhighlight %}

> `RatesHistory` 表示一个关于日元且不断增长的货币汇率 append-only 表（汇率为1）。例如，`欧元`兑`日元`从 `09:00` 到 `10:45` 的汇率为 `114`。从 `10:45` 到 `11:15`，汇率为 `116`。

Reviewer: “`RatesHistory` 表示一个关于日元且不断增长的货币汇率 append-only 表（汇率为1）” 这里的翻译意思了解，但是感觉不太通顺。

这里的意思是说这个表是相对于日元的汇率表，而且表是 append-only 的，日元对日元的汇率是 1.

> `RatesHistory` 表示一个关于日元且不断增长的货币汇率 append-only 表（汇率为1）。例如，`欧元`兑`日元`从 `09:00` 到 `10:45` 的汇率为 `114`。从 `10:45` 到 `11:15` ，汇率为 `116`。

Reviewer: 这里的翻译（这里的意思是这是一个相对于日元的汇率表，也就是说 日元对日元的 汇率是 1），感觉读起来有一点点拗口，另外 数字/字符 与中文之间需要有空格，具体可以参考 [wiki](https://cwiki.apache.org/confluence/display/FLINK/Flink+Translation+Specifications)
数字和标点符号之间不需要空格，比如 这里的 ”从 `10:45` 到 `11:15` ，“ 需要改成 ”从 `10:45` 到 `11:15`，“，也就是 5 和 ”，“ 之间不需要空格

> 假设我们要输出在 `10:58` 的时间的所有当前汇率，则需要以下 SQL 查询来计算结果表：

Reviewer: “假设我们要输出在 `10:58` 的时间的所有当前汇率" -> "假设我们要输出 `10:58` 的所有当前汇率”是否会更好一些呢

Author: 你的建议更好些，我的翻译过于直译。

  AND r2.rowtime <= TIME '10:58');
{% endhighlight %}

子查询确定对应货币的最大时间小于或等于所需时间。外部查询列出具有最大时间戳的汇率。 
 

> 下表显示了这种计算的结果。在我们的示例中， `10:58` 时表的版本，考虑了 `10:45` 时`欧元`的更新，但未考虑 `11:15` 时的`欧元`更新和`英镑`的新值。

Reviewer: 中文中很少有 “时表的版本”这种说法，这里想说的是 “10:45“ 这一时刻对应的表是啥。具体的语言可以再组织一下

> 下表显示了这种计算的结果。在我们的示例中，考虑了 `10:45` 时`欧元`的更新，但是在 `10:58` 时，该表的版本中未考虑对 `11:15` 处的`欧元`更新和`英镑`的新值。

Reviewer: “在我们的示例中，考虑了 `10:45` 时`欧元`的更新，但是在 `10:58` 时，该表的版本中未考虑对 `11:15` 处的`欧元`更新和`英镑`的新值。” -> “在我们的示例中，`10:58` 的表版本，考虑了 `10:45` 的更新，但未考虑 `11:15` 的更新” 是否会更好呢？（或者有其他更好的翻译也更好）

10:45   Euro        116
{% endhighlight %}

>  *时态表* 的概念旨在简化此类查询，加快其执行速度，并减少 Flink 的状态使用。*时态表* 是 append-only 表上的参数化视图，该视图将 append-only 表的行解释为表的变更日志，并在特定时间点提供该表的版本。将 append-only 表解释为变更日志需要指定主键属性和时间戳属性。主键确定哪些行将被覆盖，时间戳确定行有效的时间。

Reviewer: ```suggestion
 *时态表*的概念旨在简化此类查询，加快其执行速度，并减少 Flink 的状态使用。*时态表*是 append-only 表上的参数化视图，该视图将 append-only 表的行解释为表的变更日志，并在特定时间点提供该表的版本。将 append-only 表解释为变更日志需要指定主键属性和时间戳属性。主键确定哪些行将被覆盖，时间戳确定行有效的时间。
```

>  *时态表* 的概念旨在简化此类查询，加快其执行速度，并减少 Flink 的状态使用率。*时态表* 是 append-only 表上的参数化视图，该视图将 append-only 表的行解释为表的变更日志，并在特定时间点提供该表的版本。将 append-only 表解释为变更日志需要指定主键属性和时间戳属性。主键确定哪些行将被覆盖，时间戳确定行有效的时间。

Reviewer: ```suggestion
 *时态表* 的概念旨在简化此类查询，加快其执行速度，并减少 Flink 的状态使用。*时态表* 是 append-only 表上的参数化视图，该视图将 append-only 表的行解释为表的变更日志，并在特定时间点提供该表的版本。将 append-only 表解释为变更日志需要指定主键属性和时间戳属性。主键确定哪些行将被覆盖，时间戳确定行有效的时间。
```

在上面的示例中，`currency` 是 `RatesHistory` 表的主键，而 `rowtime` 是时间戳属性。

> 在Flink中, 这由[*时态表函数*](#temporal-table-function)表示.

Reviewer: ```suggestion
在 Flink 中, 这由[*时态表函数*](#temporal-table-function)表示.
```

### 与维表变化相关

另一方面，某些用例需要连接变化的维表，该表是外部数据库表。

> 假设 `LatestRates` 是一个被物化的最新汇率表（例如，存储在其中）。`LatestRates` 是物化的 `RatesHistory` 历史。那么 `LatestRates` 表在 `10:58` 的内容将是：

Reviewer: 个人感觉，这里”例如，存储在其中“ 感觉怪怪的，或许可以看看是否有更好的翻译，或者可以等其他人看是否合适

> 假设 `LatestRates` 是一个以最新汇率实现的表（例如，存储在其中）。`LatestRates` 是物化的 `RatesHistory` 历史。然后在时间 `10:58` 的 `LatestRates`表的内容将是：

Reviewer: “假设 `LatestRates` 是一个以最新汇率实现的表（例如，存储在其中）” -> “假设 `LatestRates` 是一个表示最新汇率的表（例如，存储在其中）”
“然后在时间 `10:58` 的 `LatestRates`表的内容将是：” -> “那么`LatestRates` 表在 `10:58` 的内容将是：”

是不是会更好一些呢？

Pounds      108
{% endhighlight %}

> 在Flink中，这由 [*时态表*](#temporal-table)表示.

Reviewer: ```suggestion
在Flink中，这由[*时态表*](#temporal-table)表示.
```

<a name="temporal-table-function"></a>

时态表函数
------------------------

为了访问时态表中的数据，必须传递一个[时间属性](time_attributes.html)，该属性确定将要返回的表的版本。

>  Flink 使用[表函数]({{ site.baseurl }}/zh/dev/table/functions/udfs.html#table-functions) 的SQL语法提供一种表达它的方法。

Reviewer: ```suggestion
 Flink 使用[表函数]({{ site.baseurl }}/zh/dev/table/functions/udfs.html#table-functions) 的 SQL 语法提供一种表达它的方法。
```

>  Flink 使用[表函数]({{ site.baseurl }}/zh/dev/table/functions/udfs.html#table-functions) 的 SQL 语法提供一种表达它的方法。

Reviewer: ```suggestion
 Flink 使用[表函数]({{ site.baseurl }}/zh/dev/table/functions/udfs.html#table-functions)的 SQL 语法提供一种表达它的方法。
```

09:00   Yen           1
{% endhighlight %}

对 `Rates（timeAttribute）` 的每个查询都将返回给定 `timeAttribute` 的 `Rates` 状态。

> **注意**： 当前 Flink 不支持使用常量时间属性参数直接查询时态表函数。目前，时态表函数只能在 join 中使用。上面的示例用于提供有关函数 `Rates(timeAttribute)` 返回内容的直观信息。

Reviewer: ```suggestion
**注意**：当前 Flink 不支持使用常量时间属性参数直接查询时态表函数。目前，时态表函数只能在 join 中使用。上面的示例用于提供有关函数 `Rates(timeAttribute)` 返回内容的直观信息。
```

“上面的示例用于提供有关函数 `Rates(timeAttribute)` 返回内容的直观信息。” 有一个更好的翻译吗？个人感觉意思没啥问题，读起来优点不太通顺

> 另请参阅有关[用于持续查询的 join ](joins.html)的页面，以获取有关如何与时态表 join 的更多信息。

Reviewer: ```suggestion
另请参阅有关[用于持续查询的 join ](joins.html)页面，以获取有关如何与时态表 join 的更多信息。
```

</div>
</div>

> 行 `(1)` 创建了一个 `rates` [时态表函数](#temporal-table-function),

Reviewer: 可以的话，建议把代码的注释也翻译一下

这使我们可以在[ Table API ](../tableApi.html#joins)中使用 `rates` 函数。

行 `(2)` 在表环境中注册名称为 `Rates` 的函数，这使我们可以在[ SQL ]({{ site.baseurl }}/zh/dev/table/sql/queries.html#joins)中使用 `Rates` 函数。

<a name="temporal-table"></a>

## 时态表

<span class="label label-danger">注意</span> 仅 Blink planner 支持此功能。

> 为了访问时态表中的数据，当前必须使用 `LookupableTableSource` 定义一个 `TableSource` 。Flink使用 `FOR SYSTEM_TIME AS OF` 的SQL语法查询时态表，它在SQL:2011中被提出。

Reviewer: ```suggestion
为了访问时态表中的数据，当前必须使用 `LookupableTableSource` 定义一个 `TableSource` 。Flink 使用 SQL:2011 中提出的 `FOR SYSTEM_TIME AS OF` 的 SQL 语法查询时态表。
```

> 假设我们定义了一个时态表 `LatestRates` ，我们可以通过以下方式查询此类表：

Reviewer: ```suggestion
假设我们定义了一个时态表 `LatestRates`，我们可以通过以下方式查询此类表：
```
”查询此类表“ 改成 ”查询此表“？

Yen           1
{% endhighlight %}

> **注意**： 当前，Flink 不支持以固定时间直接查询时态表。目前，时态表只能在 join 中使用。上面的示例用于提供有关时态表 `LatestRates` 返回内容的直观信息。

Reviewer: ```suggestion
**注意**：当前，Flink 不支持以固定时间直接查询时态表。目前，时态表只能在 join 中使用。上面的示例用于提供有关时态表 `LatestRates` 返回内容的直观信息。
```
